partition_table.go

```diff
@@ -194,43 +194,58 @@ func (p *PartitionTable) Close() error {
 func (p *PartitionTable) createStorage(ctx context.Context) (*storageProxy, error) {
 	var (
 		err  error
+		errs = new(multierr.Errors)
 		st   storage.Storage
 		done = make(chan struct{})
 	)
+
 	start := time.Now()
 	ticker := time.NewTicker(5 * time.Minute)
 	defer ticker.Stop()
+	defer close(done)
 	go func() {
-		defer close(done)
-		st, err = p.builder(p.topic, p.partition)
-		if err != nil {
-			return
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-done:
+				return
+			case <-ticker.C:
+				p.log.Printf("creating storage for topic %s/%d for %.1f minutes ...", p.topic, p.partition, time.Since(start).Minutes())
+			}
 		}
-		err = st.Open()
 	}()
 
-WaitLoop:
-	for {
-		select {
-		case <-ctx.Done():
-			return nil, nil
-		case <-ticker.C:
-			p.log.Printf("creating storage for topic %s/%d for %.1f minutes ...", p.topic, p.partition, time.Since(start).Minutes())
-		case <-done:
-			p.log.Debugf("finished building storage for topic %s/%d in %.1f minutes", p.topic, p.partition, time.Since(start).Minutes())
-			if err != nil {
-				return nil, fmt.Errorf("error building storage: %v", err)
-			}
-			break WaitLoop
+	st, err = p.builder(p.topic, p.partition)
+	if err != nil {
+		return nil, fmt.Errorf("error building storage: %v", err)
+	}
+	err = st.Open()
+	if err != nil {
+		errs.Collect(st.Close())
+		errs.Collect(fmt.Errorf("error opening storage: %v", err))
+		return nil, errs.NilOrError()
+	}
+
+	// close the db if context was cancelled before the builder returned
+	select {
```

Comment:
The reason we implemented it this way was to be able to shutdown the processor while starting, because sometimes the loading could take a long when leveldb does compaction on loading.
But you're right that it will fail on rebalancing before the load is completed. So I guess this solution is correct but might lead to container kills when the container management cannot cleanly shutdown.

Reply:
If I understand correctly, you are referring to the case where this function takes long enough to force a container manager to kill the container. In that case, it's explicit that the shutdown was not clean, rather than hiding this failure and causing others.

Reply:
yep, that's what I meant. But you've got a point there and we should do it this way.